File: Snap Sockets/CSharpClient/message.cs
using System;
using System.Text;

namespace CSharpClient
{
    public class message
    {
        public static readonly int header_length = 4;
        public static readonly int id_length = 4;
        public static readonly int max_body_length = 200000;
        public static readonly int username_length = 24;

        public message()
        {
            _body_length = 0;
        }

        public message( string buf )
        {
            write( buf );
        }

        public byte[] data()
        {
            return _data;
        }

        public int id_offset()
        {
            return header_length;
        }

        public int body_offset()
        {
            return header_length + id_length;
        }

        public int username_offset()
        {
            return header_length + id_length + _body_length;
        }

        public void body_length( int len )
        {
            _body_length = len > max_body_length ? max_body_length : len;
        }

        public int total_length( bool username_attached )
        {
            if( username_attached )
                return header_length + id_length + _body_length + username_length;
            else
                return header_length + id_length + _body_length;
        }

        public void write( string buf )
        {
            body_length( buf.Length );
            for( int i = 0; i < _body_length && i < buf.Length; ++i )
                _data[body_offset() + i] = Convert.ToByte( ( buf[i] ) );
            encode_header();
        }

        public bool decode_header()
        {
            _body_length = BitConverter.ToInt32( _data, 0 );
            return _body_length >= 0;
        }

        private void encode_header()
        {
            unchecked //IIIIIIIIIII'M A BAAAAD PERSOOOOOONNN
            {
                _data[3] = (byte)( ( _body_length >> 24 ) & 0xFF );
                _data[2] = (byte)( ( _body_length >> 16 ) & 0xFF );
                _data[1] = (byte)( ( _body_length >> 8 ) & 0xFF );
                _data[0] = (byte)( ( _body_length ) & 0xFF );
            }
        }

        public bool decode_id()
        {
            _id = BitConverter.ToInt32( _data, id_offset() );

            return true;
        }

        public void encode_id( int id )
        {
            unchecked //IIIIIIIIIII'M A BAAAAD PERSOOOOOONNN
            {
                _data[3 + id_offset()] = (byte)( ( id >> 24 ) & 0xFF );
                _data[2 + id_offset()] = (byte)( ( id >> 16 ) & 0xFF );
                _data[1 + id_offset()] = (byte)( ( id >> 8 ) & 0xFF );
                _data[0 + id_offset()] = (byte)( ( id ) & 0xFF );
            }
        }

        public bool decode_username()
        {
            byte[] temp = new byte[username_length];
            Array.Copy( _data, username_offset(), temp, 0, username_length );
            _username = Convert.ToString( temp );
            return true;
        }

        public void encode_username( string username )
        {
            _username = username;
            for( int idx = 0; idx < username.Length && idx < username_length; idx++ )
            {
                _data[username_offset() + idx] = Convert.ToByte( username[idx] );
            }
        }

        public override string ToString()
        {
            return Encoding.ASCII.GetString( _data, message.header_length + message.id_length, _body_length );
        }

        public byte[] _data { get; private set; } = new byte[header_length + id_length + max_body_length + username_length];
        public int _body_length { get; private set; }
        public int _id { get; private set; }
        string _username;
    }
}

File: Snap Sockets/SharedCode/message.cpp
/*****************************************************************
* Author: Jacob Asmuth
* Project: SharedCode
* Filename: message.cpp
* Date Created: 12/2/2015
* Modifications:
*
****************************************************************/
#include "message.h"

//Constructor message()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
message::message(): _body_length( 0 ), _username( "" ) {}

//Constructor message(const char * buf)
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
message::message( const char * buf )
{
	write( buf );
}

//char* data()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
char * message::data()
{
	return _data;
}

//char* id()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
char * message::id()
{
	return _data + header_length;
}

//char* body()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
char* message::body()
{
	return _data + header_length + id_length;
}

//char* username()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
char * message::username()
{
	return _data + header_length + id_length + _body_length;
}

//size_t total_length()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
size_t message::total_length() const
{
#ifdef CLIENT
		return header_length + id_length + _body_length + username_length;
#elif SERVER
		return header_length + id_length + _body_length;
#endif
}

// body_length()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
size_t message::body_length() const
{
	return _body_length;
}

// body_length(size_t new_length)
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
void message::body_length( size_t new_length )
{
	_body_length = new_length > max_body_length ? max_body_length : new_length;
}

// write(const char * buf)
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
void message::write( const char * buf )
{
	body_length( strlen( buf ) );
	memcpy( body(), buf, _body_length );
	encode_header();
}

// decode_header()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
//when we decode it just take the bytes and convert them to an int. gotta love pointer gymnastics.
bool message::decode_header()
{
	_body_length = *(int *)_data;
	return _body_length >= 0;
}

// encode_header()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
//I know this looks like hell. Basically just write 8 bits at a time to the buffer
void message::encode_header()
{
	_data[3] = ( _body_length >> 24 ) & 0xFF;
	_data[2] = ( _body_length >> 16 ) & 0xFF;
	_data[1] = ( _body_length >> 8 ) & 0xFF;
	_data[0] = _body_length & 0xFF ;
}

// decode_id()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
void message::decode_id()
{
	_id = *(int*)id();
}

// encode_id(int id)
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
void message::encode_id( int id )
{
	this->id()[3] = ( id >> 24 ) & 0xFF;
	this->id()[2] = ( id >> 16 ) & 0xFF;
	this->id()[1] = ( id >> 8 ) & 0xFF;
	this->id()[0] = id & 0xFF;
}

// decode_username()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
void message::decode_username()
{
	char buf[message::username_length] = { '\0' };
	for( unsigned i = 0; i < int( message::username_length ); ++i )
		buf[i] = username()[i];
	_username = buf;
}

// encode_username(string username)
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
void message::encode_username( string username )
{
	memset( this->username(), '\0', message::username_length );
	for( unsigned i = 0; i <= (int)message::username_length && i < username.length(); ++i )
		this->username()[i] = username[i];
}

// get_id()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
int message::get_id() const
{
	return _id;
}

// get_username()
/*****************************************************************
* Purpose:
*
* Entry:
*
* Exit:
*
****************************************************************/
string message::get_username() const
{
	return _username;
}
